### ISSUE TITLE
$_TRIALS as an object

### ISSUE THREAD

**Issue by gikeymarcia** [gikeymarcia/Collector#120]
Instead of having the plain array of `$_SESSION['Trials']` we should make a class that contains the `$_SESSION['Trials']` data but has a lot of nice methods to make array navigation easy.

````php
$_SESSION['Trials'] = new experimentController();
$_TRIALS =& $_SESSION['Trials'];
````
Proposed methods:
* `Advance()` : add one to the position counter
* `LastUse()`: return the trial contents from the last time this stimuli was used. Return `false` if this is the first time we are using the item
* `LastUseOfType($trialType)`: return the trial contents from the last time this stimuli was used as the given `$trialType`. Return `false` if this is the first time we are using the item as the given trial type. For example, `$_TRIALS->LastUseOfType('mcpic');`
* `alias()`: does what create alias does
* `current()`: returns the array contents of the current position
* `toCurrent($array)`: instead of modifying the array directly we can pull `current()` do what we want to it and then send the modified version of the trial back `toCurrent()`
* `splice()`: insert a new trial after the current trial (by default). If given a parameter it will specify how much further ahead the splice should occur. It will automatically shift anything after the splice up by one position.
* `pos(#)`: give me the information about a given trial
* `posRel(#)`: give me information about a trial relative to the current trial. You can pass it positive numbers for looking ahead (e.g., `posRel(2)` gives me 2 trials ahead) or negative numbers for looking back (e.g., `posRel(-2)` gives me 2 trials back).
* `lastResponse()`: return the `['Response']['Response']` from the last time this item was used and generated a response.
* `lastAcc($type)`: `$type` can be `strict` or `len` and will return the last trial that used this stimuli item and had either a `strictAcc` or `lenientAcc` set
* `lastRT()`: return the RT from the last time this stimuli item was used.

**Comment by adamblake** [comment]
What should `alias` do, in more or less general terms?

**Comment by TysonKerr** [comment]
Return an associative array for the php function `extract()` to use. The current solution can only create these aliases in the global scope, but if we instead camelcase all the keys and return the array, we can create the aliases inside whatever scope we want. So, something like:

```php
    function prepareForExtract(array $array) {
        $output = array();
        
        foreach ($array as $key => $value) {
            $key = preg_replace('/[^0-9a-zA-Z_ ]/', '', $key);
            $key = camelCase($key);
            if ($key === '') continue;
            if (is_numeric($key[0])) $key = '_' . $key;
            $output[$key] = $value;
        }
        
        return $output;
    }
```
and then we can use this on the current procedure row and stimuli row, and do something like
```php
$currentStim = $_EXP->getCurrentStimuli();
extract(prepareForExtract($currentStim), EXTR_PREFIX_SAME, 'stim_');
$currentProc = $_EXP->getCurrentProcedure();
extract(prepareForExtract($currentProc), EXTR_PREFIX_SAME, 'proc_');
```

**Comment by adamblake** [comment]
Why do you remove the first number, letter, underscore, or space from the key?

**Comment by adamblake** [comment]
Also, if I'm understanding what we want to use this for correctly, I think I have already implemented this in a different way. Currently I have this function:
```php
    /**
     * Sets or updates data in the trial at the given position.
     * @param int $pos The position of the trial in the experiment.
     * @param array $data variable => contents pairs to set to the trial.
     * @param string $prefix Prefixes all inserted keys with the given string 
     *                       and an underscore.
     */
    public function setTrialData($pos, array $data, $prefix = null)
    {
        foreach($data as $key => $val) {
            if ($prefix) {
                $key = $prefix . '_' . $key;
            }
            $trial[$key] = $val;
        }
    }
```
This function was written with the idea in mind that the `$expt->getCurrentTrial()` will return an array of the variables available to the trial which are then passed around to various places. For example like,
```php
$trial = $expt->getCurrentTrial();

// do stuff, setup html

<div class='stim'>
    <?php show($trial['cue']); ?>
</div>
```
Functions like the one above and `setCurrentTrialData()` will insert `[$variable => $contents]' into the trial.

**Comment by TysonKerr** [comment]
do you mean this line?
```php
            $key = preg_replace('/[^0-9a-zA-Z_ ]/', '', $key);
```
The goal of this line is to remove any non-letter/number/underscore from the name. So, a column like "Terrible {column} name% here$$$$" would becomes "Terrible column name here".

With your code, the `extract()` would be easier, just
```php
extract(prepareForExtract($trial));
```

**Comment by adamblake** [comment]
Ah, my fault! ---forgot that when the caret is in the brackets it stands for negation, not "start of line"

I'll add that method in as well (to prepare the vars)

**Comment by adamblake** [comment]
Fixed some bugs: e3967a48378714b295a0dd04d641c08a65f3fab4
Changed private methods/props to protected for extensibility: 5dd225b5a6d10364496a5f5831a02fc7b020bc6d

**Comment by adamblake** [comment]
1a19371cf8af66d5d56b68c8a08fc31467afce4f Adds phpunit config and phpunit tests for Experiment Class

**Comment by TysonKerr** [comment]
One method that might be helpful is a way to skip trials. A lot of times, people explain their procedure to me as "Ask Question A. If they answer "Yes", then do the next question, else skip the next question." I usually build some trial type that just does both questions in the same trial, but I think for other people the more intuitive thing would be to make two trials with an option to skip the second one. It would be cool if it could be used to skip multiple trials, in a way that doesn't require lots of page refreshes (i.e., not just advancing the trial and using a header refresh to try again, only to find that this trial is skipped also). And, I think people would usually want the output to say that this trial was skipped, in some additional column, perhaps. So, maybe at the top of the trial type (or maybe in the helper file somewhere), we could have something like

scoring file
```php
if ($lenientAcc > 0) {
    $_SESSION['ItemsToSkip'][] = $item;
}
```
and in the display
```php
while (in_array($item, $_SESSION['ItemsToSkip'])) {
    $recordSkip = true;
    $_EXP->skipTrial($recordSkip);
    $_EXP->advance();
    $_EXP->rerunTrial();
}
```
I'm not really sure how rerunTrial() would work, but basically I was thinking of something that just lets to restart the current page (maybe have an output buffer that we can discard)


Also, how were you imagining that post trials would work? When I was thinking of prepareForExtract(), I was imagining that the specific columns had already been selected, and the "Post X " prefix had already been trimmed out. Here is some code that could be added to prepareForExtract(), but it relies on knowing which columns are specifically from the procedure file, since it doesn't apply to the stim columns

```php
// stuff not necessarily in this function
// just need the $procRow to work with, can be passed in
$position  = $this->position;
$postTrial = $this->postTrial; // some int
$trialData = $this->getTrial($position);
$procRow   = $trialData['Procedure'];

// relevant code to prepareForExtract()
$currentProcFields = array();

if ($postTrial == 0) {
    foreach ($procRow as $column => $field) {
        $colStart = substr($column, 0, 5);
        $colStart = strtolower($colStart);
        
        if ($colStart !== 'post ') {
            $currentProcFields[$column] = $field;
        }
    }
} else {
    $colPrefix = 'Post ' . $postTrial . ' ';
    $colPrefix = strtolower($colPrefix);
    $colPreLen = strlen($colPrefix);
    
    foreach ($procRow as $column => $field) {
        $colStart = substr($column, 0, $colPreLen);
        $colStart = strtolower($colStart);
        if ($colStart === $colPrefix) {
            $cleanColumn = substr($column, $colPreLen);
            $cleanColumn = trim($cleanColumn);
            
            $currentProcFields[$cleanColumn] = $field;
        }
    }
    
    // always just use the "Item" column, and not "Post 1 Item"
    $currentProcFields['Item'] = $procRow['Item'];
}

return $currentProcFields;
```

On a different note, could `setTrialData()` somehow detect casing differences? So, if I tell it to change the "notes", and I'm currently using the "Post 1 Notes" column, can we make sure that we are correctly modifying the procedure column "Post 1 Notes"?

**Comment by TysonKerr** [comment]
Also, this is just a question about good practices, but is the reason you have `camelCase()` and `rangeToArray()` imported as nested code in `prepareForExtract()` or `parseList()` just to make this class something that can operate independently? When I made pathfinder, I debated whether or not it should use the `fileExists()` function to return a path with the real casing, but I ultimately decided not to combine the two.

**Comment by adamblake** [comment]
For the last comment here:
The `camelCase` was only actually in there the first time because I copied your comment and worked off of it. So, it was just a mistake and didn't run until I actually wrote the code for it. I don't think I used anything called `rangeToArray` --- is that a function you have somewhere?

In any case, from our last talk before I started writing the class I think you mentioned something like 'scrap it and start clean,'  so my method was just to start writing tests out to describe the behaviors the class should be capable of and then I wrote the code for it to pass the tests. We might want refactor out some stuff in the future that is shared with other classes, but for now I was just writing it to operate independently.
